File: 198-house-robber/house-robber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    if(nums.length === 1){
        return nums[0];
    }
    if(nums.length === 2){
        return Math.max(nums[0], nums[1]);
    }
    let prev = nums[0];
    let lastBest = Math.max(nums[0], nums[1]);
    for(let x = 2; x < nums.length; x++){
        let prevBest = lastBest;
        lastBest = Math.max(lastBest, prev + nums[x]);
        prev = prevBest;
    }
    return lastBest;
};

File: 198-house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if(len(nums) == 0):
            return 0
        if(len(nums) == 1):
            return nums[0]
        if(len(nums) == 2):
            return max(nums[0], nums[1])

        prev = nums[0]
        curr = max(nums[0], nums[1])

        for x in range(2, len(nums)):
            newCurr = max(curr, prev + nums[x])
            prev = curr
            curr = newCurr

        return curr